### PR TITLE
`ref readonly` is not allowed in local_variable_initializer

### LINKED ISSUE #828: C# 7 §13.6.2  `ref readonly` in local_variable_initializer
**Describe the bug**

In the C# 7 draft, §13.6.2 (Statements / Declaration statements / Local variable declarations) specifies the following restriction for "implicitly typed local variable declaration":

> The *local_variable_initializer* shall be an expression, optionally preceded by `ref` or `ref readonly`.

However, `ref readonly` at the start of a *local_variable_initializer* is not allowed by the grammar and not allowed by Roslyn either.  The grammar allows `ref readonly` in *local_variable_declaration* in front of *local_variable_type*, but not in front of an expression.

**Example**

The text of the draft gives the impression that `ref readonly i` is valid in the following.  It is not.

```csharp
public class C {
    static readonly int i;
    
    void M()
    {
        ref readonly var huh = ref readonly i;
    }
}
```

**Expected behavior**

Say just "optionally preceded by `ref`" and don't mention `ref readonly`.

**Additional context**

[Declaring and using ref readonly locals](https://github.com/dotnet/csharplang/blob/cd336064a26a43c31c1164ef7cd3f5feb4420d20/proposals/csharp-7.2/readonly-ref.md#declaring-and-using-ref-readonly-locals) in the C# 7.2 proposal shows the initialization `ref readonly var r1 = ref M1();` without using `ref readonly` in the initializer.

### PR DESCRIPTION
**Pull request by pkulikov**
Fixes #828

### CODE REVIEW

File: standard/statements.md [hunk 1]
```diff
@@ -335,9 +335,9 @@ In the context of a local variable declaration, the identifier `var` acts as a c
 
 - The *local_variable_declaration* cannot include multiple *local_variable_declarator*s.
 - The *local_variable_declarator* shall include a *local_variable_initializer*.
-- The *local_variable_initializer* shall be an *expression*, optionally preceded by `ref` or `ref readonly`.
+- The *local_variable_initializer* shall be an *expression*, optionally preceded by `ref`.
 - The initializer *expression* shall have a compile-time type.
-- The initializer *expression* cannot refer to the declared variable itself
+- The initializer *expression* cannot refer to the declared variable itself.
```
Comment (KalleOlaviNiemitalo) [inline standard/statements.md:340]: This reminds me about `var x = nameof(x);` which is currently invalid as discussed in <https://github.com/dotnet/roslyn/issues/766> and <https://github.com/dotnet/csharplang/blob/cd336064a26a43c31c1164ef7cd3f5feb4420d20/meetings/2015/LDM-2015-03-04.md>.

As the language design team did not approve relaxing that restriction for `nameof`, it should not be changed in the C# 7 standard either; but a note could be added.

Continued in <https://github.com/dotnet/csharpstandard/issues/832>

Comment (pkulikov) [reply standard/statements.md:340]: Resolving it, as #832 is tracking this.

**Review by KalleOlaviNiemitalo — APPROVED** [review]

**Comment by jskeet** [comment]
@Nigel-Ecma and @BillWagner : I think this sounds right - any concerns?

**Comment by Nigel-Ecma** [comment]
@jskeet & @BillWagner – This isn’t a right/wrong issue, both the current and proposed text are valid designs. Requiring `readonly` looks like intentional design to me – which makes Roslyn non-compliant; I was hoping someone who *knew* the designer's intention would weigh in. If the intention is to not allow/require `readonly` then this PR is good.

**Comment by KalleOlaviNiemitalo** [comment]
Generally, C# allows `ref readonly` only in front of a type, not in front of an expression.  For example, the `return` statement and the `? :` operator allow `ref` but not `ref readonly`.  If `ref readonly` were allowed (or even required) in *local_variable_initializer*, that would be particularly inconsistent with ref reassignment in C# 7.3.

Possible combinations of `ref` and `ref readonly`:

```csharp
public class C {
    int a = 0;
    readonly int b = 0;

    public void M() {
        ref int ra1 = ref a;
        ref int ra2 = ref readonly a;
        ref readonly int ra3 = ref a;
        ref readonly int ra4 = ref readonly a;
        
        ref int rb1 = ref b;
        ref int rb2 = ref readonly b;
        ref readonly int rb3 = ref b;
        ref readonly int rb4 = ref readonly b;
    }
}
```

In the Roslyn implementation, the initializations of `ra2`, `ra4`, `rb2`, and `rb4` are invalid because `ref readonly` is not allowed in the initializer; and the initialization of `rb1` is invalid because you cannot make a mutable reference to `readonly int b`.

If `ref readonly` were allowed and required in the initializer, then I think:

- `ra1` should be OK.
- `ra2` should be invalid (cannot convert `ref readonly int` to `ref int`).
- `ra3` should be OK (implicit conversion from `ref int` to `ref readonly int`).
- `ra4` should be OK.
- `rb1` should be invalid (cannot reference `readonly int b` with plain `ref`).
- `rb2` should be invalid (cannot convert `ref readonly int` to `ref int`).
- `rb3` should be invalid (cannot reference `readonly int b` with plain `ref`).
- `rb4` should be OK.

And the standard would need extra wording to achieve that.

**Comment by KalleOlaviNiemitalo** [comment]
[Declaring and using ref readonly locals](https://github.com/dotnet/csharplang/blob/cd336064a26a43c31c1164ef7cd3f5feb4420d20/proposals/csharp-7.2/readonly-ref.md#declaring-and-using-ref-readonly-locals) in the C# 7.2 proposal shows the initialization `ref readonly var r1 = ref M1();` without using `ref readonly` in the initializer.

**Comment by Nigel-Ecma** [comment]
@KalleOlaviNiemitalo – It could be said that your `rb1` example shows C# as being less clear than it could be; that said either design is valid. I would wager that this PR follows the intended design, not least because `= ref` is classified as an operator – though whether that was intended by the original design or was just a choice of the compiler team in this particular case I can’t say for certain… ;-)

**Comment by KalleOlaviNiemitalo** [comment]
The difference between `ref` and the address-of operator `&` is most clear in `? :`.

```csharp

public class C {
    public unsafe void M() {
        int x = 1, y = 2, z = 3;
        int i = 0;
        ref int r = ref i == 0 ? ref x : ref (i == 1 ? ref y : ref z);
        int* p = i == 0 ? &x : (i == 1 ? &y : &z);
    }
}
```

[SharpLab](https://sharplab.io/#v2:CYLg1APgAgzABFATHAwnA3gWAFBzw+AVwDsBnAQwDMBTBAFjgFkAKASgx3y7gEtiAXOAA84AXjgBGADRwAnmLiIZALwUwA3J254+gngoAMm3NrgAnapV4DzCi1f2jxBuAH5zl4XBAerzR+ISbr5y3iHKrMam1vwAVHAADgoBcC7uAGQiPv5igcHp8j7pEVH4AL44ZUA=)

The subexpression (i == 1 ? ref y : ref z) evaluates to a reference but it still needs `ref` in front of it in order to preserve the referenceness through the outer `? :` expression.

**Comment by KalleOlaviNiemitalo** [comment]
> your rb1 example shows C# as being less clear than it could be

Because rb1 would be invalid in both designs, I don't understand your point.

**Comment by Nigel-Ecma** [comment]
@KalleOlaviNiemitalo – I completely confused you as to the point I was making. Using an underscore to indicate a strong binding the compiler  parses your first assignment example as:

```
ref int ra1 =_ref a;
```

i.e. `ref` is not a unary operator applied to an *variable_reference* which impacts how it is treated; `= ref` is a distinct operator from `=`; and something similar applies to the ref-conditional.

Is this binding a language design decision or a compiler implementation choice? As I said I’d wager which one it is but I don’t *know*, and I don't *know* where `= ref readonly` came from and its not illogical *per se*.

Let's wait for someone who *knows*.

**Comment by KalleOlaviNiemitalo** [comment]
`= ref readonly` is not in the proposal [Readonly references](https://github.com/dotnet/csharplang/blob/cd336064a26a43c31c1164ef7cd3f5feb4420d20/proposals/csharp-7.2/readonly-ref.md).  I guess it was added by mistake when the C# standard draft was edited.  Or did the editors use some other source than this proposal?

**Comment by BillWagner** [comment]
I *think* the intention is to disallow `ref readonly` in the variable initializer (which makes this PR OK.)

Tagging @cston, @rikkigibson and @jaredpar who would know for certain.

**Comment by KalleOlaviNiemitalo** [comment]
<https://github.com/dotnet/roslyn/issues/118> "Proposal: Ref Returns and Locals" (from Jan 28, 2015) includes this example:

```csharp
public static readonly ref int Max(
    readonly ref int first, readonly ref int second, readonly ref int third)
{
    readonly ref int max = first > second ? ref first : ref second;
    return max > third ? ref max : ref third;
}
```

which differs from the current implementation in three ways: (1) `readonly ref` instead of `ref readonly`; (2) initializes a ref local with `=` rather than `= ref`; (3) returns a ref with `return` rather than `return ref`.  So, these aspects were apparently changed after that was posted.

Ref locals were mentioned in the following meeting notes and discussion issues, but those changes were not:

- [C# Design Meeting Notes for Jan 21, 2015](https://github.com/dotnet/csharplang/blob/cd336064a26a43c31c1164ef7cd3f5feb4420d20/meetings/2015/LDM-2015-01-21.md), discussion <https://github.com/dotnet/roslyn/issues/98>
- [C# Design Meeting Notes for Mar 24, 2015](https://github.com/dotnet/csharplang/blob/cd336064a26a43c31c1164ef7cd3f5feb4420d20/meetings/2015/LDM-2015-03-24.md), discussion <https://github.com/dotnet/roslyn/issues/1898>
- [C# Language Design Review, Mar 25, 2015](https://github.com/dotnet/csharplang/blob/cd336064a26a43c31c1164ef7cd3f5feb4420d20/meetings/2015/LDM-2015-03-25-Design-Review.md) (very briefly), discussion <https://github.com/dotnet/roslyn/issues/1921>

[C# Design Meeting Sep 1 2015](https://github.com/dotnet/csharplang/blob/main/meetings/2015/LDM-2015-09-01.md) shows the `= ref` initialization syntax and `return ref`.  The syntax for ref assignment and "readonly refs" had not been decided yet.

More brief mentions of ref locals or readonly refs:

- [C# Language Design Notes for Nov 1, 2016](https://github.com/dotnet/csharplang/blob/cd336064a26a43c31c1164ef7cd3f5feb4420d20/meetings/2016/LDM-2016-11-01.md)
- [C# Language Design Notes for Jan 11, 2017](https://github.com/dotnet/csharplang/blob/cd336064a26a43c31c1164ef7cd3f5feb4420d20/meetings/2017/LDM-2017-01-11.md)

[C# Language Design Notes for Feb 22, 2017](https://github.com/dotnet/csharplang/blob/cd336064a26a43c31c1164ef7cd3f5feb4420d20/meetings/2017/LDM-2017-02-22.md) links to the `ref readonly` proposal <https://github.com/dotnet/csharplang/issues/38>.  Particularly <https://github.com/dotnet/csharplang/issues/38#issuecomment-279844069> ponders the differences between `ref readonly`, `readonly ref`, and `readonly ref readonly`.  Doesn't seem to have discussion about `= ref` grammar.

[C# Language Design Notes for Feb 28, 2017](https://github.com/dotnet/csharplang/blob/cd336064a26a43c31c1164ef7cd3f5feb4420d20/meetings/2017/LDM-2017-02-28.md) introduces "Conditional operator over refs" <https://github.com/dotnet/csharplang/issues/223> and discusses how many `ref` keywords the syntax should require.  **There is a decision not to use `condition ? ref readonly v1 : ref readonly v2`, and a note that "`ref` is not part of the expression".**

[C# Language Design Notes for Mar 1, 2017](https://github.com/dotnet/csharplang/blob/cd336064a26a43c31c1164ef7cd3f5feb4420d20/meetings/2017/LDM-2017-03-01.md) continues with conditional refs and approves the `condition ? ref v1 : ref v2` syntax.

[C# Language Design Notes for May 31, 2017](https://github.com/dotnet/csharplang/blob/cd336064a26a43c31c1164ef7cd3f5feb4420d20/meetings/2017/LDM-2017-05-31.md) discusses `modreq` generated from some `ref readonly`.  Doesn't discuss C# syntax for refs.

[C# Language Design Notes for Jul 5, 2017](https://github.com/dotnet/csharplang/blob/cd336064a26a43c31c1164ef7cd3f5feb4420d20/meetings/2017/LDM-2017-07-05.md) approves shipping the ref features in C# 7.2.

[C# Language Design Notes for Aug 28, 2017](https://github.com/dotnet/csharplang/blob/cd336064a26a43c31c1164ef7cd3f5feb4420d20/meetings/2017/LDM-2017-08-28.md) discusses ref-like types.

[C# Language Design Notes for Sep 25. 2017](https://github.com/dotnet/csharplang/blob/cd336064a26a43c31c1164ef7cd3f5feb4420d20/meetings/2017/LDM-2017-09-25.md) adds `ref readonly` locals.  **This shows the initialization `ref readonly var r = ref a[1];` but also considers shorter syntax `ref readonly int x = a[1];` near the end.**

[C# Language Design Notes for Sep 27, 2017](https://github.com/dotnet/csharplang/blob/cd336064a26a43c31c1164ef7cd3f5feb4420d20/meetings/2017/LDM-2017-09-27.md) discusses argument syntax for `ref readonly` parameters, and decides to require `ref` in front of the argument expression but not `ref readonly`.  This also introduces `in` parameters.

[Milestone philosophy](https://github.com/dotnet/csharplang/blob/cd336064a26a43c31c1164ef7cd3f5feb4420d20/meetings/2017/LDM-2017-10-02.md) (Oct 2, 2017) again shows `ref readonly tmp = ref Get();` syntax.

I think this shows that the language design team decided to initialize with `= ref` and not `= ref readonly`.

**Comment by jaredpar** [comment]
> I think this shows that the language design team decided to initialize with = ref and not = ref readonly.

That is correct. This also follows to more recent features like [ref readonly] where `ref` is still a legal passing mechanism. The design team decided the complexity of allowing a `ref readonly` expression syntax didn't outweigh the benefits and stuck with simply `ref`.

**Review by Nigel-Ecma — APPROVED** [review]
Unsurprisingly it has been confirmed that the “[language] design team decided the complexity of allowing a `ref readonly` expression syntax didn’t outweigh the benefits and stuck with simply `ref`”. So this is all good :-)

**Pull request merged**